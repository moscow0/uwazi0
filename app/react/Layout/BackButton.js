import PropTypes from 'prop-types';
import React from 'react';

import { Icon } from 'UI';
import { I18NLink, t } from 'app/I18N';

const BackButton = ({ to, className }) => (
  <I18NLink to={to} className={`btn btn-default ${className}`}>
    <Icon icon="arrow-left" directionAware />
    <span className="btn-label">{t('System', 'Back')}</span>
  </I18NLink>
);

BackButton.defaultProps = {
  to: '',
  className: '',
};

BackButton.propTypes = {
  to: PropTypes.string,
  className: PropTypes.string,
};

export default BackButton;
