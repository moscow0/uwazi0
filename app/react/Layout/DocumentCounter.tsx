import React from 'react';
import { Translate } from 'app/I18N';

export interface EntityCounterProps {
  selectedEntitiesCount: number;
  entityListCount: number;
  entityTotal: number;
  totalConnectionsCount?: number;
  hitsTotalRelation: string;
}

export const DocumentCounter = (props: EntityCounterProps) => {
  const {
    totalConnectionsCount,
    selectedEntitiesCount,
    entityListCount,
    entityTotal,
    hitsTotalRelation,
  } = props;
  const totalEntitiesValue = <b> {`${entityTotal}${hitsTotalRelation === 'gte' ? '+' : ''}`} </b>;
  const entityCounter = (
    <>
      {selectedEntitiesCount > 0 && (
        <>
          <b> {selectedEntitiesCount} </b> <Translate>selected of</Translate>
        </>
      )}
      <b> {entityListCount} </b> <Translate>shown of</Translate>
      {totalEntitiesValue}
      <Translate>entities</Translate>
    </>
  );
  const connectionsCounter = (
    <>
      <b>{totalConnectionsCount!} </b>
      <Translate>connections</Translate>, <b>{totalEntitiesValue} </b>
      <Translate>entities</Translate>
    </>
  );

  return <span>{totalConnectionsCount === undefined ? entityCounter : connectionsCounter}</span>;
};
