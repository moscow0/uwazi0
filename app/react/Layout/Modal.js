import PropTypes from 'prop-types';
import React, { Component } from 'react';
import ReactModal from 'react-modal';
import { Translate } from 'app/I18N';

export default class Modal extends Component {
  render() {
    const style = {
      overlay: { zIndex: this.props.zIndex, backgroundColor: 'rgba(0, 0, 0, 0.75)' },
    };
    const { type, className, children, isOpen } = this.props;
    return (
      <ReactModal
        style={style}
        className={`modal-dialog modal-${type} ${className}`}
        isOpen={isOpen}
        contentLabel=""
        ariaHideApp={false}
      >
        <div className="modal-content">{children}</div>
      </ReactModal>
    );
  }
}

const childrenType = PropTypes.oneOfType([PropTypes.object, PropTypes.array, PropTypes.string]);

Modal.propTypes = {
  isOpen: PropTypes.bool,
  type: PropTypes.string,
  children: childrenType,
  className: PropTypes.string,
  zIndex: PropTypes.number,
};
Modal.defaultProps = {
  isOpen: false,
  type: 'success',
  className: '',
  zIndex: 100,
};
const Body = ({ children }) => <div className="modal-body">{children}</div>;
Body.propTypes = { children: childrenType };

const Header = ({ children }) => <div className="modal-header">{children}</div>;
Header.propTypes = { children: childrenType };

const Footer = ({ children }) => <div className="modal-footer">{children}</div>;
Footer.propTypes = { children: childrenType };

const Title = ({ children }) => <h4 className="modal-title">{children}</h4>;
Title.propTypes = { children: childrenType };

const Close = ({ onClick }) => (
  <button type="button" className="close" onClick={onClick}>
    <span aria-hidden="true">&times;</span>
    <span className="sr-only">
      <Translate>Close</Translate>
    </span>
  </button>
);
Close.propTypes = { onClick: PropTypes.func };

Modal.Body = Body;
Modal.Header = Header;
Modal.Footer = Footer;
Modal.Title = Title;
Modal.Close = Close;
