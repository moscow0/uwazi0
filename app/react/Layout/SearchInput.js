import React, { Component } from 'react';
import { Icon } from 'UI';
import { t } from 'app/I18N';
import ModalTips from 'app/App/ModalTips';
import { SearchTipsContent } from 'app/App/SearchTipsContent';

export class SearchInput extends Component {
  render() {
    return (
      <div className="input-group">
        <input type="text" className="form-control" placeholder="Search" {...this.props} />
        <Icon icon="search" />
        <ModalTips
          label={t('System', 'Search Tips', null, false)}
          title={t('System', 'Narrow down your searches', null, false)}
        >
          <SearchTipsContent />
        </ModalTips>
      </div>
    );
  }
}

export default SearchInput;
