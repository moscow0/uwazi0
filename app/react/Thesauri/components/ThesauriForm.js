/* eslint-disable max-lines */
import ShowIf from 'app/App/ShowIf';
import FormGroup from 'app/DocumentForm/components/FormGroup';
import { BackButton } from 'app/Layout';
import { DragAndDropContainer } from 'app/Layout/DragAndDrop';
import { notEmpty } from 'app/Metadata/helpers/validator';
import {
  addGroup,
  addValue,
  importThesaurus,
  removeValue,
  saveThesaurus,
  sortValues,
  updateValues,
} from 'app/Thesauri/actions/thesauriActions';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { actions as formActions, Field, Form } from 'react-redux-form';
import { bindActionCreators } from 'redux';
import { Icon } from 'UI';
import { I18NLink, Translate } from 'app/I18N';

import { SelectFileButton } from 'app/App/SelectFileButton';
import { ThesauriFormItem } from './ThesauriFormItem';

function sanitizeThesauri(thesaurus) {
  const sanitizedThesauri = { ...thesaurus };
  sanitizedThesauri.values = sanitizedThesauri.values
    .filter(value => value.label)
    .filter(value => !value.values || value.values.length)
    .map(value => {
      const _value = { ...value };
      if (_value.values) {
        _value.values = _value.values.filter(_v => _v.label);
      }
      return _value;
    });
  return sanitizedThesauri;
}

class ThesauriForm extends Component {
  static validation(thesauris, id) {
    return {
      name: {
        duplicated: val =>
          !thesauris.find(
            thesauri =>
              thesauri.type !== 'template' &&
              thesauri._id !== id &&
              thesauri.name.trim().toLowerCase() === val.trim().toLowerCase()
          ),
        required: notEmpty,
      },
    };
  }

  constructor(props) {
    super(props);
    this.save = this.save.bind(this);
    this.renderItem = this.renderItem.bind(this);
    this.onChange = this.onChange.bind(this);
    this.importThesaurusFile = this.importThesaurusFile.bind(this);
  }

  componentDidMount() {
    this.firstLoad = true;
  }

  componentDidUpdate(previousProps) {
    if (this.firstLoad) {
      this.firstLoad = false;
      return;
    }

    this.addEmptyValueAtTheEnd();

    this.focusIfWasGroup(previousProps);
  }

  componentWillUnmount() {
    this.props.resetForm('thesauri.data');
    this.props.setInitial('thesauri.data');
  }

  onChange(values, groupIndex) {
    this.props.updateValues(values, groupIndex);
  }

  importThesaurusFile(file) {
    const thes = sanitizeThesauri(this.props.thesauri);
    this.props.importThesaurus(thes, file);
  }

  focusIfWasGroup(previousProps) {
    const { values } = this.props.thesauri;
    const previousValues = previousProps.thesauri.values;
    const addedValue = values.length > previousProps.thesauri.values.length;
    const lastValueIsGroup = values.length && values[values.length - 1].values;
    const previousLastValueWasGroup =
      previousValues.length && previousValues[previousValues.length - 1].values;
    if (lastValueIsGroup && (!previousLastValueWasGroup || addedValue)) {
      this.groups[this.groups.length - 1].focus();
    }
  }

  addEmptyValueAtTheEnd() {
    this.props.thesauri.values.forEach((value, index) => {
      if (
        value.values &&
        (!value.values.length || value.values[value.values.length - 1].label !== '')
      ) {
        this.props.addValue(index);
      }
    });

    if (
      !this.props.thesauri.values.length ||
      this.props.thesauri.values[this.props.thesauri.values.length - 1].label !== ''
    ) {
      this.props.addValue();
    }
  }

  save(thesauri) {
    const sanitizedThesauri = sanitizeThesauri(thesauri);
    this.props.saveThesaurus(sanitizedThesauri);
  }

  renderItem(value, index) {
    return (
      <ThesauriFormItem
        ref={f => this.groups.push(f)}
        value={value}
        index={index}
        removeValue={this.props.removeValue}
        onChange={this.onChange}
      />
    );
  }

  render() {
    const isNew = this.props.new;
    const id = this.props.thesauri._id;
    const { values } = this.props.thesauri;
    if (!isNew && !id) {
      return false;
    }

    this.groups = [];
    return (
      <div className="thesauri">
        <Form
          model="thesauri.data"
          onSubmit={this.save}
          validators={ThesauriForm.validation(this.props.thesauris.toJS(), this.props.thesauri._id)}
        >
          <div className="panel panel-default thesauri">
            <div className="panel-heading">
              <FormGroup {...this.props.state.name} submitFailed={this.props.state.submitFailed}>
                <Field model=".name">
                  <input
                    id="thesauriName"
                    className="form-control"
                    type="text"
                    placeholder="Thesauri name"
                  />
                  <ShowIf
                    if={
                      this.props.state.$form.touched &&
                      this.props.state.name &&
                      this.props.state.name.errors.duplicated
                    }
                  >
                    <div className="validation-error">
                      <Icon icon="exclamation-triangle" size="xs" />
                      &nbsp;
                      <Translate>Duplicated name</Translate>
                    </div>
                  </ShowIf>
                </Field>
              </FormGroup>
            </div>
            <div className="thesauri-values">
              <div className="">
                <b>
                  <Translate>Items:</Translate>
                </b>
              </div>
              <DragAndDropContainer
                onChange={this.onChange}
                renderItem={this.renderItem}
                items={values}
                iconHandle
              />
            </div>
            <div className="settings-footer">
              <div className="btn-cluster">
                <BackButton to="/settings/dictionaries" className="btn-plain" />
              </div>
              <div className="btn-cluster lg-margin-left-12 sm-order-1 sm-footer-extra-row">
                <button type="button" className="btn btn-default" onClick={this.props.addGroup}>
                  <Icon icon="plus" />
                  <span className="btn-label">
                    <Translate>Add group</Translate>
                  </span>
                </button>
                <button type="button" className="btn btn-default" onClick={this.props.sortValues}>
                  <Icon icon="sort-alpha-down" />
                  <span className="btn-label">
                    <Translate>Sort</Translate>
                  </span>
                </button>
                <SelectFileButton onFileImported={this.importThesaurusFile}>
                  <button type="button" className="btn btn-default import-template">
                    <Icon icon="upload" />
                    <span className="btn-label">
                      <Translate>Import</Translate>
                    </span>
                  </button>
                </SelectFileButton>
              </div>
              <div className="btn-cluster content-right">
                <I18NLink to="/settings/dictionaries" className="btn btn-default btn-extra-padding">
                  <span className="btn-label">
                    <Translate>Cancel</Translate>
                  </span>
                </I18NLink>
                <button type="submit" className="btn btn-success btn-extra-padding save-template">
                  <span className="btn-label">
                    <Translate>Save</Translate>
                  </span>
                </button>
              </div>
            </div>
          </div>
        </Form>
      </div>
    );
  }
}

ThesauriForm.defaultProps = {
  new: false,
};

ThesauriForm.propTypes = {
  resetForm: PropTypes.func.isRequired,
  setInitial: PropTypes.func.isRequired,
  saveThesaurus: PropTypes.func.isRequired,
  addValue: PropTypes.func.isRequired,
  addGroup: PropTypes.func.isRequired,
  sortValues: PropTypes.func.isRequired,
  removeValue: PropTypes.func.isRequired,
  updateValues: PropTypes.func.isRequired,
  importThesaurus: PropTypes.func.isRequired,
  thesauris: PropTypes.object.isRequired,
  thesauri: PropTypes.object.isRequired,
  state: PropTypes.object.isRequired,
  new: PropTypes.bool,
};

function mapStateToProps(state) {
  return {
    thesauri: state.thesauri.data,
    thesauris: state.thesauris,
    state: state.thesauri.formState,
  };
}

function bindActions(dispatch) {
  return bindActionCreators(
    {
      saveThesaurus,
      importThesaurus,
      addValue,
      addGroup,
      sortValues,
      removeValue,
      updateValues,
      resetForm: formActions.reset,
      setInitial: formActions.setInitial,
      validate: formActions.validate,
    },
    dispatch
  );
}

export { ThesauriForm, mapStateToProps };

export default connect(mapStateToProps, bindActions, null, { withRef: true })(ThesauriForm);
