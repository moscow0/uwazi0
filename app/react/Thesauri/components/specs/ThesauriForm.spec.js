/** @format */

import React from 'react';
import { shallow } from 'enzyme';
import Immutable from 'immutable';
import { DragAndDropContainer } from 'app/Layout/DragAndDrop';
import { mapStateToProps, ThesauriForm } from 'app/Thesauri/components/ThesauriForm.js';

describe('ThesauriForm', () => {
  let props;
  let component;
  let instance;
  beforeEach(() => {
    props = {
      thesauri: {
        _id: '123',
        name: 'thesauri name',
        values: [
          { label: 'Heroes', values: [{ label: 'Batman' }, { label: 'Robin' }] },
          { label: 'Villains', values: [{ label: 'Joker' }, { label: 'Penguin' }] },
        ],
      },
      resetForm: jasmine.createSpy('resetForm'),
      setInitial: jasmine.createSpy('setInitial'),
      thesauris: Immutable.fromJS([{ name: 'Countries' }]),
      saveThesaurus: jasmine.createSpy('saveThesaurus'),
      importThesaurus: jasmine.createSpy('importThesaurus'),
      addValue: jasmine.createSpy('addValue'),
      addGroup: jasmine.createSpy('addGroup'),
      sortValues: jasmine.createSpy('sortValues'),
      removeValue: jasmine.createSpy('removeValue'),
      updateValues: jasmine.createSpy('updateValues'),
      state: { fields: [], $form: {} },
    };
  });

  const render = () => {
    component = shallow(<ThesauriForm {...props} />);
    instance = component.instance();
  };

  describe('when unmount', () => {
    it('should reset the form', () => {
      render();
      component.unmount();
      expect(props.resetForm).toHaveBeenCalled();
      expect(props.setInitial).toHaveBeenCalled();
    });
  });

  describe('render', () => {
    it('should render DragAndDropContainer with thesauri items', () => {
      render();
      expect(component.find(DragAndDropContainer).props().renderItem).toBe(
        component.instance().renderItem
      );
      expect(component).toMatchSnapshot();
    });
  });

  describe('renderItem', () => {
    it('should render ThesauriFormItem with specified value and index', () => {
      render();
      const renderedItem = component.instance().renderItem(props.thesauri.values[1], 1);
      expect(renderedItem).toMatchSnapshot();
    });
  });

  describe('save()', () => {
    it('should sanitize the thesauri and save it', () => {
      render();
      const thesauri = {
        values: [
          { label: 'Heroes', values: [{ label: 'Batman' }, { label: 'Robin' }], id: 0 },
          { label: '', values: [{ label: 'Joker' }, { label: 'Penguin' }], id: 1 },
          { label: 'I have no values', values: [], id: 3 },
          { label: '', id: 4 },
        ],
      };
      instance.save(thesauri);
      const sanitizedThesauri = {
        values: [{ label: 'Heroes', values: [{ label: 'Batman' }, { label: 'Robin' }], id: 0 }],
      };
      expect(props.saveThesaurus).toHaveBeenCalledWith(sanitizedThesauri);
    });
  });

  describe('on props update', () => {
    it('should add an empty value at the end of the thesauri and at the end of groups when there is none', () => {
      render();
      instance.firstLoad = false;
      instance.componentDidUpdate(props);

      expect(props.addValue.calls.allArgs()).toEqual([[0], [1], []]);
    });
  });

  describe('componentDidUpdate() when a group is added', () => {
    it('should focus it', () => {
      render();
      props.thesauri = {
        _id: '123',
        name: 'thesauri name',
        values: [{ label: 'Heroes', values: [{ label: 'Batman' }, { label: 'Robin' }] }],
      };
      instance.firstLoad = false;
      instance.groups = [
        { focus: jasmine.createSpy('focus') },
        { focus: jasmine.createSpy('focus') },
      ];
      instance.componentDidUpdate(props);
      expect(instance.groups[1].focus).toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    let state;
    beforeEach(() => {
      state = {
        thesauri: {
          data: { name: 'thesauri name', values: [] },
          state: 'thesauri form state',
        },
        thesauris: Immutable.fromJS([{ name: 'Countries' }]),
      };
    });

    it('should map the thesauri to initialValues', () => {
      expect(mapStateToProps(state).thesauri).toEqual({ name: 'thesauri name', values: [] });
      expect(mapStateToProps(state).thesauris).toEqual(Immutable.fromJS([{ name: 'Countries' }]));
    });
  });

  describe('onChange', () => {
    it('should update values', () => {
      render();
      const values = [{ label: 'item' }];
      instance.onChange(values, 1);
      expect(props.updateValues).toHaveBeenCalledWith(values, 1);
    });
  });

  describe('validation', () => {
    describe('name duplicated', () => {
      let thesauris;
      let id;
      beforeEach(() => {
        thesauris = [
          { _id: 'id1', name: 'Countries' },
          { _id: 'id2', name: 'Cities' },
          { _id: 'id3', name: 'People', type: 'template' },
        ];
        id = 'id1';
      });
      function testValidationResult(inputValue, expectedResult) {
        const {
          name: { duplicated },
        } = ThesauriForm.validation(thesauris, id);
        const res = duplicated(inputValue);
        expect(res).toBe(expectedResult);
      }
      it('should return false if another thesaurus exists with the same name', () => {
        testValidationResult('Cities', false);
      });
      it('should return true if thesaurus with similar name is itself', () => {
        testValidationResult('Countries', true);
      });
      it('should return true if template has same name', () => {
        testValidationResult('People', true);
      });
    });
  });
});
