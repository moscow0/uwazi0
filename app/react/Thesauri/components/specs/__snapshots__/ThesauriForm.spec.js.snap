// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThesauriForm render should render DragAndDropContainer with thesauri items 1`] = `
<div
  className="thesauri"
>
  <Connect(Form)
    model="thesauri.data"
    onSubmit={[Function]}
    validators={
      Object {
        "name": Object {
          "duplicated": [Function],
          "required": [Function],
        },
      }
    }
  >
    <div
      className="panel panel-default thesauri"
    >
      <div
        className="panel-heading"
      >
        <FormGroup
          className=""
        >
          <Modeled(Connect(Field))
            model=".name"
          >
            <input
              className="form-control"
              id="thesauriName"
              placeholder="Thesauri name"
              type="text"
            />
            <ShowIf>
              <div
                className="validation-error"
              >
                <Connect(Icon)
                  icon="exclamation-triangle"
                  size="xs"
                />
                Â 
                <Connect(Translate)>
                  Duplicated name
                </Connect(Translate)>
              </div>
            </ShowIf>
          </Modeled(Connect(Field))>
        </FormGroup>
      </div>
      <div
        className="thesauri-values"
      >
        <div
          className=""
        >
          <b>
            <Connect(Translate)>
              Items:
            </Connect(Translate)>
          </b>
        </div>
        <DragDropContext(DropTarget(DragAndDropContainer))
          iconHandle={true}
          items={
            Array [
              Object {
                "label": "Heroes",
                "values": Array [
                  Object {
                    "label": "Batman",
                  },
                  Object {
                    "label": "Robin",
                  },
                ],
              },
              Object {
                "label": "Villains",
                "values": Array [
                  Object {
                    "label": "Joker",
                  },
                  Object {
                    "label": "Penguin",
                  },
                ],
              },
            ]
          }
          onChange={[Function]}
          renderItem={[Function]}
        />
      </div>
      <div
        className="settings-footer"
      >
        <div
          className="btn-cluster"
        >
          <BackButton
            className="btn-plain"
            to="/settings/dictionaries"
          />
        </div>
        <div
          className="btn-cluster lg-margin-left-12 sm-order-1 sm-footer-extra-row"
        >
          <button
            className="btn btn-default"
            onClick={[Function]}
            type="button"
          >
            <Connect(Icon)
              icon="plus"
            />
            <span
              className="btn-label"
            >
              <Connect(Translate)>
                Add group
              </Connect(Translate)>
            </span>
          </button>
          <button
            className="btn btn-default"
            onClick={[Function]}
            type="button"
          >
            <Connect(Icon)
              icon="sort-alpha-down"
            />
            <span
              className="btn-label"
            >
              <Connect(Translate)>
                Sort
              </Connect(Translate)>
            </span>
          </button>
          <SelectFileButton
            onFileImported={[Function]}
          >
            <button
              className="btn btn-default import-template"
              type="button"
            >
              <Connect(Icon)
                icon="upload"
              />
              <span
                className="btn-label"
              >
                <Connect(Translate)>
                  Import
                </Connect(Translate)>
              </span>
            </button>
          </SelectFileButton>
        </div>
        <div
          className="btn-cluster content-right"
        >
          <Connect(I18NLink)
            className="btn btn-default btn-extra-padding"
            to="/settings/dictionaries"
          >
            <span
              className="btn-label"
            >
              <Connect(Translate)>
                Cancel
              </Connect(Translate)>
            </span>
          </Connect(I18NLink)>
          <button
            className="btn btn-success btn-extra-padding save-template"
            type="submit"
          >
            <span
              className="btn-label"
            >
              <Connect(Translate)>
                Save
              </Connect(Translate)>
            </span>
          </button>
        </div>
      </div>
    </div>
  </Connect(Form)>
</div>
`;

exports[`ThesauriForm renderItem should render ThesauriFormItem with specified value and index 1`] = `
<ThesauriFormItem
  index={1}
  onChange={[Function]}
  removeValue={[Function]}
  value={
    Object {
      "label": "Villains",
      "values": Array [
        Object {
          "label": "Joker",
        },
        Object {
          "label": "Penguin",
        },
      ],
    }
  }
/>
`;
